add metadata validation for hive format

nvliyuan opened the pull request:
this fix #172 sub task: metadata validation for hive format
Signed-off-by: liyuan <yuali@nvidia.com>

nartal1 commented on data_validation/src/spark_rapids_validation_tool/validation_scripts/metadata_validation.py:
```diff
@@ -0,0 +1,273 @@
+    verify_DF = table_DF.select(verify_column)
+
+    for c in verify_DF.schema.fields:
+        # here only excluded 'data' because it will raise exception, we also should excluded str/map/nested
```
Chanfe 'data' to 'date'. 
Is there a follow on issue for excluding string/map/nested ? Curious why we are excluding string?

nvliyuan replied:
I am not sure about which should be excluded for now, maybe we can keep it there and wait for the requirement from customers.

nartal1 commented on data_validation/src/spark_rapids_validation_tool/validation_scripts/metadata_validation.py:
```diff
@@ -0,0 +1,273 @@
+    table_metric_df2 = generate_metric_df(spark, table2_DF, i, e, t2)
+    joined_table = table_metric_df1.alias("t1").join(table_metric_df2.alias("t2"), ["ColumnName"])
+
+    cond = (round("t1.min" + t1, p) != round("t2.min" + t2, p)) | \
```
Wondering how difficult would it be to extend to other data types(string, map) with this approach? Should we have a different `condition` for string and then append that to result_table?

nvliyuan replied:
will draft a follow up pr to extend to other data types(map/struct), this still depends on customers requirements.

nvliyuan replied:
add supported for str and map/nested

nartal1 reviewed (commented):
Please update copyright year to 2023 for all new files.

nvliyuan commented on data_validation/src/spark_rapids_validation_tool/csp/__init__.py:
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
```
updated all copyright

nvliyuan commented on data_validation/src/spark_rapids_validation_tool/build.py:
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
```
updated all copyright

nvliyuan commented on data_validation/docs/validation-tools.md:
````diff
@@ -0,0 +1,62 @@
+
+- Install `spark_rapids_validation_tool` with python [3.8, 3.10] using:
+   ```bash
+    git clone https://github.com/nvliyuan/spark-rapids-tools.git
````
updated

nvliyuan commented on data_validation/docs/validation-tools.md:
```diff
@@ -0,0 +1,62 @@
+# Validation tools on Dataproc
+
+This is a guide for the Validation tools for Apache Spark on [Google Cloud Dataproc](https://cloud.google.com/dataproc).  
+At the end of this guide, the user will be able to run the validation tools to analysis
```
updated

nvliyuan commented on data_validation/src/spark_rapids_validation_tool/validation_scripts/metadata_validation.py:
```diff
@@ -0,0 +1,273 @@
+    if not valid_metadata_included_column(spark, args):
+        return False
+    if args.format != 'hive':
+        print('|--Currently only support hive format--|')
```
Will draft another pr for not hive format.

nvliyuan replied:
oh, I got it, updated, thanks

nartal1 commented on data_validation/src/spark_rapids_validation_tool/validation_scripts/metadata_validation.py:
```diff
@@ -0,0 +1,263 @@
+            sql = f'select * from {table_name}'
+            if any(cond != 'None' for cond in [partition, f]):
+                where_clause = ' where ' + ' and '.join(x for x in [partition, f] if x != 'None')
+            sql += where_clause
```
This should be under `if` condition. It would throw an error if both parition and f is None because where_clause will not be set.

amahussein commented on data_validation/src/spark_rapids_validation_tool/data_validation.py:
```diff
@@ -0,0 +1,80 @@
+
+        return wrapper
+
+    def run_spark_submit(self, options, capture='all'):
+        """Run spark application via spark-submit command."""
+        cmd = ['$SPARK_HOME/bin/spark-submit']
+        cmd += options
+        stdout, stderr = self.run_cmd(cmd, capture=capture)
+        return stdout + stderr
```
IIUC, this will run the `spark-submit` through SSH to the dataproc cluster.
This means, that the tool fails if a user has no SSH access to the cluster.

If this is one of the assumption of the feature, it has to be stated explicitly to grant the SSH access .

amahussein commented on data_validation/src/spark_rapids_validation_tool/validation_scripts/metadata_validation.py:
```diff
@@ -0,0 +1,289 @@
+    The different metadata of each column in each table(min/max/avg/stddev/count_distinct):
+    (If the values are identical, then a specific cell is empty, aka NULL. So we only show differences),
+    the result dataframe should be:
+    +----------+------+------+------+------+-----------------+-----------------+-------------------+-------------------+-----------+-----------+
+    |ColumnName| min_A| min_B| max_A| max_B|            avg_A|            avg_B|           stddev_A|           stddev_B|countdist_A|countdist_B|
+    +----------+------+------+------+------+-----------------+-----------------+-------------------+-------------------+-----------+-----------+
+    |      col1|      |      |12.000|11.000|6.090909090909091|              6.0|  3.477198454346414|    3.3166247903554|           |           |
+    |      col3|      |      |      |      |5.090909090909091|5.181818181818182| 3.1766191290283903|  3.060005941764879|           |           |
+    |      col8|12.340|12.330|      |      |       12.4345455|       12.4336364|0.30031195901474267|0.30064173786328213|          3|          4|
+    +----------+------+------+------+------+-----------------+-----------------+-------------------+-------------------+-----------+-----------+
```
nit: showing a sample demo, it will be nice to make it more readable by rounding the float digits to 2 decimal points. Wide lines will fail the pylint test anyway.

amahussein reviewed (commented):
I have some general comments about the PR:

**Usability**:

- The tool seems to rely on running spark-submit through SSH cmd. This is problematic for users who has no access to SSH. In production, most of the CSP users do not have permissions to manage clusters (no SSH/create/delete). They may only have limited privileges to submit jobs and access certain resources. Is there a technical reason why the `metadata_validation.py` is not submitted as a spark job to the argument-cluster?
- Very tailored to a specific requirement and it is unlikely to fit for general purpose.

**Maintenance**:

Is there a technical reason of what drove the decision to implement that in python module?
I find that the validation logic is more intuitive to be added to the scala-core-module (scala jar).
A new class should be added to the scala core module DataValidation first. Then the wrapper can be implemented to call that class on different CSP.

_Why do I have concerns with this being in Python?_

In summary, it is the high maintenance and overhead incurred with python module.

- We need to provide the wrapper with the supported data types. We can create a static config file for that but this means that every time the plugin upgrades the support, the static file has to be updated manually and also test that the validation output is correct. That's not feasible across two repositories.
- The python wrapper is supposed to be more like an interface to hide complexity of integrating across different CSPs. It should not deal with "Spark" core. The scala-core-module is the place where we had those kind of functionalities are implemented.
- We will have to deal with different Spark/Hive versions in the python module. We just started looking into the Shimming for Core module; but with features like this PR, we will have to handle different Spark/Hive implementations Python too.
- More difficult to test: Currently, we did not set up the testing platform for the wrapper. What makes testing more challenging in the pytools is that the integration with the CSP command interface. On the other hand, having DataValidation in scala-core-module allows us to have unit tests easier and that can be re-executed efficiently. 

**Styling**: 

Variable names and arguments need to be meaningful. Arguments/variable-names should represent the purpose so that the code becomes readable. Please make sure that the code complies to this rule. I will make sure the `.pylintrc` detects such coding style.

amahussein commented:
> this fix https://github.com/NVIDIA/spark-rapids-tools/issues/172 sub task: metadata validation for hive format
Signed-off-by: liyuan [yuali@nvidia.com](mailto:yuali@nvidia.com)

@nvliyuan , a reminder that using the word "Fix issue" tells Github to close the issue automatically with the merge. 
This means that entire 172 will be closed.
The right way to do that is to create a separate issue for each individual subtask.

nartal1 commented on data_validation/src/spark_rapids_validation_tool/validation_scripts/metadata_validation.py:
```diff
@@ -0,0 +1,289 @@
+                                       when(col("t1.countDistinct"+t1) != col("t2.countDistinct"+t2), col("t1.countDistinct"+t1)).otherwise('').alias("countdist_A"),
+                                       when(col("t1.countDistinct"+t1) != col("t2.countDistinct"+t2), col("t2.countDistinct"+t2)).otherwise('').alias("countdist_B")
+                                       ).where(cond).sort(asc("ColumnName"))
+    return result_table
```
Can we add some basic stats so that it could be used for performance testing. I think this query is going to take longer time to complete.

nvliyuan commented:
Hi Teams, since we will refine to a better structure, so I commit all my code(including metadata validation and dataset validation) to a new pr #184, #184 contains all the changes and all addressed comments of this #173 pr, could you help review and maybe we could move to that pr and close this one
@nartal1 @amahussein
